[#598] improvement(api): introduce UpdateComment in SchemaChange interface to update comment of schema

SteNicholas opened the pull request:
### What changes were proposed in this pull request?

Introduce `UpdateComment` in `SchemaChange` interface and implement `UpdateComment` interface in `HiveCatalogOperations` and `IcebergCatalogOperations`.

### Why are the changes needed?

There is no entrance for updating schema comment currently, because there is no `updateComment` implementation of interface`SchemaChange` mentioned in #587.

Fix: #598

### Does this PR introduce _any_ user-facing change?

Introduce `UpdateComment` in `SchemaChange` interface to update comment of schema.

### How was this patch tested?

- `TestSchemaChange`
  - `TestSchemaChange#testUpdateComment`, `TestSchemaChange#testUpdateCommentEqualsAndHashCode` and `TestSchemaChange#testUpdateCommentNotEqualsAndHashCode`
- `TestRelationalCatalog#testUpdateSchemaComment`
- `TestSchemaOperations#testAlterSchema`
- `CatalogHiveIT#testAlterSchema`
- `CatalogIcebergIT#testOperationIcebergSchema`

xunliu commented on api/src/main/java/com/datastrato/gravitino/rel/SchemaChange.java:
```diff
@@ -47,6 +60,22 @@ static SchemaChange removeProperty(String property) {
+  @EqualsAndHashCode
+  @Getter
+  final class UpdateComment implements SchemaChange {
+    private final String newComment;
```
Maybe it's better to use `comment` for variable names?

SteNicholas replied:
@xunliu, I refers to variable name of `TableChange.UpdateComment` which also uses `newComment ` as variable name.

xunliu reviewed (commented):
hi @SteNicholas Thank you for your contribution.

SteNicholas commented:
@FANNG1, @yuqi1129, could you help to review this improvement?

mchades commented on catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:
```diff
@@ -284,7 +284,9 @@ public HiveSchema alterSchema(NameIdentifier ident, SchemaChange... changes)
       for (SchemaChange change : changes) {
-        if (change instanceof SchemaChange.SetProperty) {
+        if (change instanceof SchemaChange.UpdateComment) {
+          properties.put(COMMENT, ((SchemaChange.UpdateComment) change).getNewComment());
```
You should use `database.setDescription` to change the Hive database's comment, instead of update the property

FANNG1 commented on catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java:
```diff
@@ -246,8 +246,11 @@ public IcebergSchema alterSchema(NameIdentifier ident, SchemaChange... changes)
       for (SchemaChange change : changes) {
-        if (change instanceof SchemaChange.SetProperty) {
+        if (change instanceof SchemaChange.UpdateComment) {
+          comment = ((SchemaChange.UpdateComment) change).getNewComment();
```
I think you we should store `comment` in  Iceberg schema properties like hive. or we couldn't load the latest comment in `loadSchema`

FANNG1 replied:
@Clearvive  please  check the Iceberg related logics.

mchades commented:
Can you please add related catalog integration testing in package `package com.datastrato.gravitino.integration.test.catalog`?

SteNicholas commented:
@mchades, @FANNG1, thanks for detailed review. I have addressed above comment and added catalog integration testing via `CatalogHiveIT#testAlterSchema` and `CatalogIcebergIT#testOperationIcebergSchema`.  PTAL.

jerryshao commented:
@mchades would you please help to review?

mchades commented on common/src/main/java/com/datastrato/gravitino/dto/requests/SchemaUpdateRequest.java:
```diff
@@ -30,6 +33,35 @@ public interface SchemaUpdateRequest extends RESTRequest {
+    @Override
+    public void validate() throws IllegalArgumentException {
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(newComment),
```
Is it sufficient to check for null here, or is an empty string  `" "` also not allowed?

SteNicholas replied:
@mchades, the `UpdateTableColumnCommentRequest` validation behavior also validates whether `newComment` is blank.

mchades replied:
I mean we should allow users setting schema comment to `" "`

mchades commented:
BTW, as this PR introduces a new feature, we should update the corresponding documentation. 

Would you like to address this as part of this PR, or would you prefer to open a separate issue to track this matter? I believe both options are viable, and it is entirely up to you.

FANNG1 commented on catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:
```diff
@@ -283,8 +283,14 @@ public HiveSchema alterSchema(NameIdentifier ident, SchemaChange... changes)
+        if (change instanceof SchemaChange.UpdateComment) {
+          commentUpdated = true;
+          comment = ((SchemaChange.UpdateComment) change).getNewComment();
+          properties.put(COMMENT, comment);
```
seems there're no need to put `COMMENT` in properties since we could store schema comments in the description.
```
        alteredDatabase.setDescription(comment);
```

FANNG1 commented on catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java:
```diff
@@ -246,8 +246,14 @@ public IcebergSchema alterSchema(NameIdentifier ident, SchemaChange... changes)
+        if (change instanceof SchemaChange.UpdateComment) {
+          commentUpdated = true;
+          comment = ((SchemaChange.UpdateComment) change).getNewComment();
+          updates.put("comment", comment);
```
suggest to use `IcebergSchemaPropertiesMetadata.COMMENT` not `comment`

FANNG1 commented on catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java:
```diff
@@ -246,8 +246,14 @@ public IcebergSchema alterSchema(NameIdentifier ident, SchemaChange... changes)
       List<String> removals = new ArrayList<>();
       Map<String, String> updates = new HashMap<>();
       Map<String, String> resultProperties = new HashMap<>(metadata);
+      boolean commentUpdated = false;
```
can we remove `commentUpdated`?  we could use `comment == null` to justify whether comment is updated

mchades commented on api/src/main/java/com/datastrato/gravitino/rel/SchemaChange.java:
JDBC catalog also needs to be implemented

SteNicholas replied:
@mchades, `JdbcCatalogOperations` does not support `alterSchema` interface implementation. Therefore JDBC catalog does not need to be implemented.

mchades replied:
JDBC catalog does not support `alterSchema` yet

mchades reviewed (dismissed):
LGTM, thanks for your contribution and patience!

SteNicholas commented:
@mchades, @FANNG1, I'm fixing the failure of CI. Please do not merge.

SteNicholas commented:
@mchades, @FANNG1, the failure of CI for `CatalogHiveIT#testAlterSchema` and `CatalogIcebergIT#CatalogIcebergIT` is caused by the 2.3.9 version of `hive-metastore` which does not support [HIVE-8472](https://issues.apache.org/jira/browse/HIVE-8472) that supports `setDescription`. WDYT?

mchades commented:
Below is the support status of updating schema comments for various engines:

1. Hive
  Does not support in HQL, but can update by HMS (only works on version 2.2.1, 2.4.0, and 3.0.0  - [HIVE-8472](https://issues.apache.org/jira/browse/HIVE-8472))
2. MySQL 
  Does not support (see [MySQL doc](https://dev.mysql.com/doc/refman/8.0/en/alter-database.html))
3. PostgreSQL
  Does not support (see [PostgreSQL](https://www.postgresql.org/docs/12/ddl-schemas.html)) 
4. Iceberg
  namespace does support comment (from [source code](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/catalog/SupportsNamespaces.java))

So, most engines do not support updating schema comments. I'm not sure if Gravitino needs this feature. Can you help evaluate this? @jerryshao  @FANNG1

SteNicholas commented:
> Below is the support status of updating schema comments for various engines:
> 
> 1. Hive
>    Does not support in HQL([Hive doc](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Create/Drop/Alter/UseDatabase)), but can update by HMS (only works on version 2.2.1, 2.4.0, and 3.0.0  - [HIVE-8472](https://issues.apache.org/jira/browse/HIVE-8472))
> 2. MySQL
>    Does not support (see [MySQL doc](https://dev.mysql.com/doc/refman/8.0/en/alter-database.html))
> 3. PostgreSQL
>    Does not support (see [PostgreSQL](https://www.postgresql.org/docs/12/ddl-schemas.html))
> 4. Iceberg
>    namespace does support comment (from [source code](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/catalog/SupportsNamespaces.java))
> 
> So, most engines do not support updating schema comments. I'm not sure if Gravitino needs this feature. Can you help evaluate this? @jerryshao @FANNG1

@jerryshao, @FANNG1, `HiveCatalogOperations` and `IcebergCatalogOperations` both need the `setDescription` of HMS, but this version of HMS does not support. Therefore, IMO, the feature could be used since the version of HMS upgrades.

jerryshao commented:
> Below is the support status of updating schema comments for various engines:
> 
> 1. Hive
>    Does not support in HQL([Hive doc](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Create/Drop/Alter/UseDatabase)), but can update by HMS (only works on version 2.2.1, 2.4.0, and 3.0.0  - [HIVE-8472](https://issues.apache.org/jira/browse/HIVE-8472))
> 2. MySQL
>    Does not support (see [MySQL doc](https://dev.mysql.com/doc/refman/8.0/en/alter-database.html))
> 3. PostgreSQL
>    Does not support (see [PostgreSQL](https://www.postgresql.org/docs/12/ddl-schemas.html))
> 4. Iceberg
>    namespace does support comment (from [source code](https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/catalog/SupportsNamespaces.java))
> 
> So, most engines do not support updating schema comments. I'm not sure if Gravitino needs this feature. Can you help evaluate this? @jerryshao @FANNG1

@mchades can you please point me where does Iceberg support updating comment for namespace?

mchades commented:
Sorry, it is a typo, the correct is: namespace **does not** support comment @jerryshao

jerryshao commented:
Since most catalogs/connectors don't support altering comment for the schema currently, I think we can postpone this PR until it has the concrete requirement, what do you think? @SteNicholas @mchades @FANNG1

SteNicholas commented:
@jerryshao, I agree that this pull request could be postponed until it has the concrete requirement.

jerryshao commented:
Close stale PR, please reopen it at anytime if you wan to continue.
